cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityStatusStore.java

```diff
@@ -0,0 +1,236 @@
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.StoreDefinition;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * CapabilityStatusStore which takes care of reading , writing capability status and provides additional helpful methods
+ */
+public class CapabilityStatusStore implements CapabilityReader, CapabilityWriter {
+
+  private static final Gson GSON = new Gson();
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  CapabilityStatusStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  /**
+   * Return the current status for a capability. If capability is not present, throws {@link IllegalArgumentException}
+   *
+   * @param capability
+   * @return {@link CapabilityStatus}
+   */
+  public CapabilityStatus getStatus(String capability) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      Collection<Field<?>> keyField = Collections
+        .singleton(Fields.stringField(StoreDefinition.CapabilitiesStore.NAME_FIELD, capability));
+      Collection<String> returnField = Collections.singleton(StoreDefinition.CapabilitiesStore.STATUS_FIELD);
+      Optional<StructuredRow> result = capabilityTable.read(keyField, returnField);
+      return result.map(structuredRow -> CapabilityStatus
+        .valueOf(structuredRow.getString(StoreDefinition.CapabilitiesStore.STATUS_FIELD).toUpperCase())).orElse(null);
+    }, IOException.class);
+  }
+
+  @Override
+  public boolean isEnabled(String capability) throws IOException {
+    return getStatus(capability) == CapabilityStatus.ENABLED;
+  }
+
+  @Override
+  public void checkAllEnabled(Collection<String> capabilities) throws IOException, CapabilityNotAvailableException {
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      List<List<Field<?>>> multiKeys = new ArrayList<>();
+      capabilities.forEach(capability -> {
+        multiKeys
+          .add(Collections.singletonList(Fields.stringField(StoreDefinition.CapabilitiesStore.NAME_FIELD, capability)));
+      });
+      Map<String, String> capabilityMap = capabilityTable.multiRead(multiKeys).stream().
+        collect(Collectors.toMap(
+          structuredRow -> structuredRow.getString(StoreDefinition.CapabilitiesStore.NAME_FIELD),
+          structuredRow -> structuredRow.getString(StoreDefinition.CapabilitiesStore.STATUS_FIELD)
+        ));
+      for (String capability : capabilities) {
```

Comment:
Move this loop outside of the transaction as it has nothing to do with it. You can use the one that takes `TxCallable` to return the `capabilityMap` from the transaction.

Reply:
Done